### PR TITLE
[extractor/rai] General overhaul of the extractor

### PR DESCRIPTION
**Pull request by nixxo**
### Description of your *pull request* and other information

- added RaiCultura to handle a new website
- improved RaiSudtirol to handle more urls
- renamed _real_extract to _legacy_real_extarct in RaiIE as is no longer needed (kept as a fallback solution in the new _real_extract)
- removed unnecessary loop for media quality requests, no longer supported by the server
- better media server query to get maximum quality available by forcing User-Agent
- improved and simplified http formats creation
- added checks for media availability
- added thumbnails list method used by all extractors
- added various join_nonempty, traverse_obj, f-strings to simplify the code
- _TESTS cleanup
- commented a bit the code for clarity

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)

</details>

### CODE REVIEW

File: yt_dlp/extractor/rai.py [hunk 1]
```diff
@@ -35,139 +28,123 @@ class RaiBaseIE(InfoExtractor):
...
+            self.raise_no_formats('The relinker returned no media url')
 
+        # geo flag is a bit unreliable and not properly set all the time
+        geoprotection = xpath_text(relinker, './geoprotection', default='N') == 'Y'
+        if '/video_no_available.mp4' in media_url:
             # This does not imply geo restriction (e.g.
             # http://www.raisport.rai.it/dl/raiSport/media/rassegna-stampa-04a9f4bd-b563-40cf-82a6-aad3529cb4a9.html)
-            if '/video_no_available.mp4' in media_url:
-                continue
+            self.report_warning('Video not available. Likely due to geo-restriction.')
```
Comment (bashonly) [inline yt_dlp/extractor/rai.py:62]: per discussion in #5672 it may be preferable to make this `raise_geo_restricted` instead

Comment (nixxo) [reply yt_dlp/extractor/rai.py]: fixed in next commit

File: yt_dlp/extractor/rai.py [hunk 2]
```diff
@@ -35,189 +28,174 @@ class RaiBaseIE(InfoExtractor):
...
+            formats.append({
+                'url': media_url,
+                'vcodec': 'none',
+                'acodec': 'mp3',
+                'format_id': 'https-mp3',
+            })
+        elif ext == 'm3u8' or 'format=m3u8' in media_url:
+            formats.extend(self._extract_m3u8_formats(
+                media_url, video_id, 'mp4', m3u8_id='hls', fatal=False))
+            # TODO: after #5924 is merged add force_codecs={'acodec': 'mp4a', 'vcodec': 'avc1'}
```
Comment (nixxo) [inline yt_dlp/extractor/rai.py:73]: #5924

**Comment by matteocontrini** [comment]
If it's helpful, using `hbbtv` as the User-Agent (or passing it with `&forceUserAgent=hbbtv`) in the request to the relinker sometimes returns the URL of the MPEG-DASH manifest. Not always though, and it seem to use a limited set of qualities.

File: yt_dlp/extractor/rai.py [hunk 3]
```diff
@@ -592,57 +583,62 @@ class RaiIE(RaiBaseIE):
...
-            'subtitles': subtitles,
+            'subtitles': self._extract_subtitles(url, media),
             **relinker_info
         }
 
     def _real_extract(self, url):
+        content_info = self._extract_from_content_id(self._match_id(url), url)
+        return content_info or self._legacy_real_extract(url)
+
+    def _legacy_real_extract(self, url):
```
Comment (pukkandan) [inline yt_dlp/extractor/rai.py:636]: I am in favor of deleting this completely unless a usecase is known.

File: yt_dlp/extractor/rai.py [hunk 4]
```diff
@@ -35,189 +28,187 @@ class RaiBaseIE(InfoExtractor):
...
+                'format_id': 'https-mp3',
+            })
+        elif ext == 'm3u8' or 'format=m3u8' in media_url:
+            fmts = self._extract_m3u8_formats(
+                media_url, video_id, 'mp4', m3u8_id='hls', fatal=False)
+            for f in fmts:
+                if not f.get('acodec'):
+                    f['acodec'] = 'mp4a'
+                if not f.get('vcodec'):
+                    f['vcodec'] = 'avc1'
```
Comment (pukkandan) [inline yt_dlp/extractor/rai.py:77]: Isn't this sufficient instead of #5924 ?

**Review by pukkandan — APPROVED** [review]

**Comment by pukkandan** [comment]
Can merge if u r okay with my changes.

File: yt_dlp/extractor/rai.py (hunk 4 above)
Comment (nixxo) [reply yt_dlp/extractor/rai.py:77]: not really.

it does not fix #887 nor other examples provided in the #5924 discussion. It's a bit too generic to properly work and set codecs properly.

If you are not convinced by #5924 code you can scrap the force_codecs part in this PR and go along in merging the changes for now.

We can work on the changes provided in #5924 later.

File: yt_dlp/extractor/rai.py (hunk 3 above)
Comment (nixxo) [reply yt_dlp/extractor/rai.py:636]: for me it's ok to scrap the legacy method.

I just left it there as a fallback solution for maybe some weird webpages that might still exist but that I couldn't find.
But again, I'm 99% sure this is useless.

File: yt_dlp/extractor/rai.py (hunk 4 above)
Comment (pukkandan) [reply yt_dlp/extractor/rai.py:77]: Ok, it seems I'm misunderstanding the point of `force_codecs`. let me take another look at it.

Comment (nixxo) [reply yt_dlp/extractor/rai.py]: you probably understood correctly what force_codecs does, but #5924 does also extra checks on the m3u8 manifests to properly set the codecs, and so, a generic solution like this one is not always ok.

**Pull request merged**